packages/logging/src/logging-and-telemetry.ts

```diff
@@ -0,0 +1,785 @@
+} from '@mongosh/types';
+import { inspect } from 'util';
+import { MongoLogWriter } from 'mongodb-log-writer';
+import { mongoLogId } from 'mongodb-log-writer';
+import type {
+  AnalyticsIdentifyMessage,
+  MongoshAnalytics,
+  MongoshAnalyticsIdentity,
+} from './analytics-helpers';
+import { hookLogger } from '@mongodb-js/devtools-connect';
+import { MultiSet, toSnakeCase } from './helpers';
+import { Writable } from 'stream';
+
+export class MongoshLoggingAndTelemetry {
+  log: MongoLogWriter;
+  pendingLogEvents: CallableFunction[] = [];
+
+  telemetryAnonymousId: string | undefined;
+  userId: string | undefined;
+  trackingProperties: {
+    mongosh_version: string;
+    session_id: string;
+  };
+  userTraits: AnalyticsIdentifyMessage['traits'] & {
+    [key: string]: unknown;
+  };
+
+  isSetup = false;
+  hasMongoLogWriterInitialized = false;
+
+  constructor(
+    public bus: MongoshBus,
+    public analytics: MongoshAnalytics,
+    public providedTraits: { platform: string } & {
+      [key: string]: unknown;
+    },
+    mongoshVersion: string
+  ) {
+    const dummySink = new Writable({
+      write(chunk, encoding, callback) {
+        callback();
+      },
+    });
+    this.log = new MongoLogWriter('', null, dummySink);
+
+    this.userTraits = {
+      ...providedTraits,
+      session_id: this.log.logId,
+    };
+
+    this.trackingProperties = {
+      mongosh_version: mongoshVersion,
+      session_id: this.log.logId,
+    };
+  }
+
+  public setup() {
+    this._setupBusEventListeners();
+    this.isSetup = true;
+  }
+
+  public setupLogger(logger: MongoLogWriter) {
+    if (!this.isSetup) {
+      throw new Error('Run setup() before setting up the log writer.');
+    }
+    this.log = logger;
+
+    this.userTraits.session_id = this.log.logId;
+    this.trackingProperties.session_id = this.log.logId;
+  }
+
+  public getTelemetryUserIdentity = (): MongoshAnalyticsIdentity => ({
+    anonymousId: this.telemetryAnonymousId ?? (this.userId as string),
+  });
+
+  /** Start processing pending events after mongosh:logger-initialized */
+  private onLoggerInitialized() {
+    this.hasMongoLogWriterInitialized = true;
+    for (const pendingEvent of this.pendingLogEvents) {
+      pendingEvent();
+    }
+    this.pendingLogEvents = [];
+
+    hookLogger(this.bus, this.log, 'mongosh', (uri) =>
+      redactURICredentials(uri)
+    );
+  }
+
+  private _setupBusEventListeners() {
+    this.bus.on(
+      'mongosh:logger-initialized',
+      this.onLoggerInitialized.bind(this)
+    );
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const onBus = <T extends any[]>(
+      event: keyof MongoshBusEventsMap,
+      listener: (...args: T) => void
+    ) => {
+      this.bus.on(event, (...args: T) => {
+        if (this.hasMongoLogWriterInitialized) {
+          listener(...args);
+        } else {
+          this.pendingLogEvents.push(() => listener(...args));
+        }
+      });
+    };
```

Comment:
The template makes this looks like it would be type-safe, but right now it doesn't actually do anything towards that end (i.e. the `T` could as well just be `any[]` here) – here's a typesafe version of this:

```suggestion
    const onBus = <K extends keyof MongoshBusEventsMap>(
      event: K,
      listener: (...args: Parameters<MongoshBusEventsMap[K]>) => void
    ) => {
      this.bus.on(event, ((...args: Parameters<MongoshBusEventsMap[K]>) => {
        if (this.hasMongoLogWriterInitialized) {
          listener(...args);
        } else {
          this.pendingLogEvents.push(() => listener(...args));
        }
      }) as MongoshBusEventsMap[K]);
    };
```

Otherwise there's nothing that correlates the listener parameters to the actual event name (which is what's supposed to be defining the listener parameters).